#### Part I: Backwards Elimination
Use this data file, which contains test scores and IQ for 15 individuals. Each individual took 5 tests. The IQ is the response variable, and the five different tests are the potential predictor variables. Perform a backwards elimination on this data, then answer the following questions:

Which model is the best? Why?
From the best model, what is the adjusted R2 value and what does it mean?
From the best model, how does each variable influence IQ?

In [9]:
IQlist <- read.csv("C:/Users/Marcy/Documents/Entity Coursework/DS106 Machine Learning/Lesson 4 Modeling with step-wise regression/IQ.csv")

In [10]:
head(IQlist)

Test1 Test2 Test3 Test4 Test5 IQ 
1 83    34    65    63    64    106
2 73    19    73    48    82     92
3 54    81    82    65    73    102
4 96    72    91    88    94    121
5 84    53    72    68    82    102
6 86    72    63    79    57    105

#### create a model using IQ as the response variable, and the tests as predictor variables. 

In [11]:
Fitall <- lm(IQ~., data=IQlist)

In [12]:
summary(Fitall)


Call:
lm(formula = IQ ~ ., data = IQlist)

Residuals:
    Min      1Q  Median      3Q     Max 
-10.255  -5.661  -2.414   3.576  22.154 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)   
(Intercept) 85.24038   23.69513   3.597  0.00577 **
Test1       -1.93357    1.02910  -1.879  0.09297 . 
Test2       -1.65988    0.87290  -1.902  0.08966 . 
Test3        0.10495    0.21990   0.477  0.64454   
Test4        3.77838    1.83450   2.060  0.06952 . 
Test5       -0.04058    0.20122  -0.202  0.84467   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 10.65 on 9 degrees of freedom
Multiple R-squared:  0.3991,	Adjusted R-squared:  0.06522 
F-statistic: 1.195 on 5 and 9 DF,  p-value: 0.3835


#### Although the overall p value at the bottom is significant, none of the individual predictors are, meaning that the model is probably not a good fit; the number of variables has just inflated the p value to the point of being significant, and what the independent variables are does not really matter.

In [5]:
step(Fitall, direction = "backward")

Start:  AIC=75.31
IQ ~ Test1 + Test2 + Test3 + Test4 + Test5

        Df Sum of Sq    RSS    AIC
- Test5  1      4.61 1025.8 73.378
- Test3  1     25.85 1047.0 73.685
<none>               1021.2 75.310
- Test1  1    400.56 1421.7 78.274
- Test2  1    410.29 1431.5 78.376
- Test4  1    481.32 1502.5 79.103

Step:  AIC=73.38
IQ ~ Test1 + Test2 + Test3 + Test4

        Df Sum of Sq    RSS    AIC
- Test3  1     21.24 1047.0 71.685
<none>               1025.8 73.378
- Test1  1    403.17 1429.0 76.350
- Test2  1    406.47 1432.3 76.384
- Test4  1    481.26 1507.1 77.148

Step:  AIC=71.69
IQ ~ Test1 + Test2 + Test4

        Df Sum of Sq    RSS    AIC
<none>               1047.0 71.685
- Test2  1    406.25 1453.3 74.603
- Test1  1    415.39 1462.4 74.697
- Test4  1    484.13 1531.2 75.386



Call:
lm(formula = IQ ~ Test1 + Test2 + Test4, data = IQ)

Coefficients:
(Intercept)        Test1        Test2        Test4  
     90.733       -1.965       -1.649        3.789  


#### A combination of tests 1, 2 and 4 is a better model than any that omit them. 

#### but is this model any good? create a new model with just those variables and look at the summary.

In [14]:
fitsome = lm(IQ~Test1 + Test2 + Test4, data = IQlist)

In [8]:
summary (fitsome)


Call:
lm(formula = IQ ~ Test1 + Test2 + Test4, data = IQ)

Residuals:
     Min       1Q   Median       3Q      Max 
-10.9184  -6.8179  -0.9142   4.3920  21.1950 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  90.7327    12.8272   7.073 2.06e-05 ***
Test1        -1.9650     0.9406  -2.089   0.0607 .  
Test2        -1.6485     0.7980  -2.066   0.0632 .  
Test4         3.7890     1.6801   2.255   0.0455 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 9.756 on 11 degrees of freedom
Multiple R-squared:  0.3839,	Adjusted R-squared:  0.2158 
F-statistic: 2.284 on 3 and 11 DF,  p-value: 0.1356


#### With a multiple R-squared value of .38, interpretation of this means the model explais 38.3% of the variation in the IQ variable, with another 61.7% chalked up to noise or random error. The p value is larger than .05, and these two figures together suggest this isn't a very good model. The coefficient of Test1 is -1.97 and Test2 is -1.65. This means with everything else equal, a decrease on these tests shows a lower IQ. But the higher the score on Test 4, has a positive correlation with IQ scores. 

#### To interpret the coefficients, typically the coefficient of a variable is interpreted as the change in the response (the tests) based on a 1-unit change in the corresponding explanatory variable (IQ) keeping all other variables held constant.

#### Part II: Compare Stepwise Regression Types
The following dataset will be used for this analysis. This data has a single response (Y) variable, and twelve predictor (X1 through X12) variables. Use these data to run all three kinds of stepwise regressions (backward elimination, forward selection, and the hybrid method). After completing these analyses, answer the following questions:

Which model was the best for each type of method?
How do the final models from each method compare to each other?
From your chosen "best model," explain what variable(s) contribute to predicting Y and for how much variance they account.

In [15]:
data = read.csv("C:/Users/Marcy/Documents/Entity Coursework/DS106 Machine Learning/Lesson 4 Modeling with step-wise regression/stepwiseRegression.csv")

In [16]:
head(data)

Y      X1   X2    X3   X4    X5   X6    X7   X8    X9   X10  X11   X12 
1 443.87 74.4 -20.9 29.9  -9.1 -2.3  -0.3 43.3  -5.9 25.7 72.6 -40.3  2.9
2 134.11 75.8 -16.9 33.1 -65.6  0.4 -29.2 49.6  -1.3 11.7 71.5 -37.3  2.7
3 619.19 52.2 -16.9 30.8 -57.4  0.5  34.4 69.9  -5.1 17.9 73.8 -43.7 -1.6
4 465.01 25.8 -23.5 32.4 -52.5 -1.0   4.0 52.1  -9.5 11.9 72.2 -43.7 -1.3
5 609.92 52.8  -7.6 33.5 -34.3 -0.6  33.0 49.2  -1.8 33.5 74.9 -40.4  0.5
6 275.96 73.2 -10.9 27.6 -56.2 -1.6  -7.0 88.8 -16.6 26.4 74.0 -41.4  2.0

#### backward to start

In [17]:
fitsall = lm(Y~., data=data)
summary(fitsall)


Call:
lm(formula = Y ~ ., data = data)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.1410 -1.7124 -0.1732  1.5415  4.9986 

Coefficients:
              Estimate Std. Error  t value Pr(>|t|)    
(Intercept)  1.410e+03  1.496e+01   94.260   <2e-16 ***
X1          -9.897e-03  1.815e-02   -0.545   0.5865    
X2          -7.567e-02  4.301e-02   -1.759   0.0812 .  
X3          -1.007e-01  9.998e-02   -1.007   0.3160    
X4           2.810e+00  9.887e-03  284.252   <2e-16 ***
X5          -1.972e-01  1.945e-01   -1.014   0.3128    
X6           5.988e+00  7.560e-03  792.080   <2e-16 ***
X7           1.157e-02  1.033e-02    1.120   0.2651    
X8           2.921e-02  3.311e-02    0.882   0.3794    
X9          -1.738e-02  3.021e-02   -0.575   0.5661    
X10         -1.194e+01  1.886e-01  -63.302   <2e-16 ***
X11         -1.454e-01  9.145e-02   -1.590   0.1145    
X12         -2.601e+01  1.200e-01 -216.772   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1


In [19]:
step(fitsall, direction="backward")

Start:  AIC=220.74
Y ~ X1 + X2 + X3 + X4 + X5 + X6 + X7 + X8 + X9 + X10 + X11 + 
    X12

       Df Sum of Sq     RSS     AIC
- X1    1         2     588  219.07
- X9    1         2     588  219.11
- X8    1         4     590  219.61
- X3    1         5     591  219.87
- X5    1         5     591  219.88
- X7    1         6     592  220.13
<none>                  586  220.74
- X11   1        13     599  221.53
- X2    1        16     602  222.14
- X10   1     20422   21008  676.88
- X12   1    239478  240064  988.69
- X4    1    411780  412366 1057.94
- X6    1   3197390 3197977 1320.13

Step:  AIC=219.07
Y ~ X2 + X3 + X4 + X5 + X6 + X7 + X8 + X9 + X10 + X11 + X12

       Df Sum of Sq     RSS     AIC
- X9    1         2     590  217.53
- X8    1         4     591  217.88
- X3    1         5     592  218.11
- X5    1         5     593  218.15
- X7    1         6     593  218.32
<none>                  588  219.07
- X11   1        13     600  219.81
- X2    1        15     602  220.21
- 


Call:
lm(formula = Y ~ X2 + X4 + X6 + X10 + X11 + X12, data = data)

Coefficients:
(Intercept)           X2           X4           X6          X10          X11  
 1410.27177     -0.06975      2.80806      5.98660    -11.97829     -0.13102  
        X12  
  -25.98121  


#### An AIC of 213.38 using X2, X4, X6, X10-12 as the predictors yields the lowest AIC. now compare model summary with all the predictors

In [21]:
fitssomed = lm(Y~X2 + X4 + X6 +X10 +X11 + X12, data=data)
summary(fitssomed)


Call:
lm(formula = Y ~ X2 + X4 + X6 + X10 + X11 + X12, data = data)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.3190 -1.7438 -0.0897  1.7094  5.3882 

Coefficients:
              Estimate Std. Error  t value Pr(>|t|)    
(Intercept)  1.410e+03  1.320e+01  106.815   <2e-16 ***
X2          -6.975e-02  4.158e-02   -1.677   0.0961 .  
X4           2.808e+00  9.434e-03  297.644   <2e-16 ***
X6           5.987e+00  7.410e-03  807.924   <2e-16 ***
X10         -1.198e+01  1.742e-01  -68.769   <2e-16 ***
X11         -1.310e-01  8.950e-02   -1.464   0.1458    
X12         -2.598e+01  1.171e-01 -221.937   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.241 on 121 degrees of freedom
Multiple R-squared:  0.9999,	Adjusted R-squared:  0.9998 
F-statistic: 1.385e+05 on 6 and 121 DF,  p-value: < 2.2e-16


#### Wow, the multiple R-squared value is .99, which means this model explains 99% of the variation in the Y variable! That's great. But let's try forward selection next.

In [24]:
forward1 = lm(Y~1, data=data)
summary(forward1)


Call:
lm(formula = Y ~ 1, data = data)

Residuals:
    Min      1Q  Median      3Q     Max 
-481.25 -129.04   -5.43  138.24  519.84 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   427.61      16.03   26.68   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 181.3 on 127 degrees of freedom


### Note the intercept is 427.61, which is just the mean of Y in the dataset. 

In [25]:
step(forward1, direction = "forward", scope=formula(fitsall))

Start:  AIC=1332.26
Y ~ 1

       Df Sum of Sq     RSS    AIC
+ X6    1   3312523  862636 1132.4
+ X12   1    565679 3609480 1315.6
+ X4    1    340596 3834563 1323.4
+ X2    1     88438 4086721 1331.5
+ X5    1     77171 4097988 1331.9
<none>              4175159 1332.3
+ X9    1     39558 4135601 1333.0
+ X7    1     39324 4135835 1333.0
+ X1    1     15448 4159711 1333.8
+ X3    1     10545 4164613 1333.9
+ X10   1     10472 4164687 1333.9
+ X11   1      4176 4170982 1334.1
+ X8    1       741 4174418 1334.2

Step:  AIC=1132.41
Y ~ X6

       Df Sum of Sq    RSS     AIC
+ X4    1    584861 277775  989.36
+ X12   1    392692 469944 1056.67
+ X10   1     29736 832900 1129.92
+ X8    1     14958 847678 1132.17
<none>              862636 1132.41
+ X3    1      9152 853483 1133.05
+ X9    1      5341 857295 1133.62
+ X11   1      5119 857517 1133.65
+ X2    1       904 861732 1134.28
+ X1    1       378 862258 1134.36
+ X5    1        63 862573 1134.40
+ X7    1        36 862600 1134.41



Call:
lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = data)

Coefficients:
(Intercept)           X6           X4          X12          X10           X2  
 1410.27177      5.98660      2.80806    -25.98121    -11.97829     -0.06975  
        X11  
   -0.13102  


#### The final iteration brings us to   AIC=213.38, which includes
Y ~ X6 + X4 + X12 + X10 + X2 + X11    -- It happens these are the same from the backward elimination model.

#### build a model that contains only these

In [26]:
finalmodelfwd = lm(Y ~ X6 + X4 + X12 + X10 + X2 + X11, data=data)
summary (finalmodelfwd)


Call:
lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = data)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.3190 -1.7438 -0.0897  1.7094  5.3882 

Coefficients:
              Estimate Std. Error  t value Pr(>|t|)    
(Intercept)  1.410e+03  1.320e+01  106.815   <2e-16 ***
X6           5.987e+00  7.410e-03  807.924   <2e-16 ***
X4           2.808e+00  9.434e-03  297.644   <2e-16 ***
X12         -2.598e+01  1.171e-01 -221.937   <2e-16 ***
X10         -1.198e+01  1.742e-01  -68.769   <2e-16 ***
X2          -6.975e-02  4.158e-02   -1.677   0.0961 .  
X11         -1.310e-01  8.950e-02   -1.464   0.1458    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.241 on 121 degrees of freedom
Multiple R-squared:  0.9999,	Adjusted R-squared:  0.9998 
F-statistic: 1.385e+05 on 6 and 121 DF,  p-value: < 2.2e-16


#### The multiple R-squared here too is .9999 - The AIC for the forward model is 213.38, whereas the AIC for the backwards was also exactly 213.38.  Handson directions say to check the hybrid, so that's next up. 

In [27]:
step(forward1, direction = "both", scope=formula(fitsall))

Start:  AIC=1332.26
Y ~ 1

       Df Sum of Sq     RSS    AIC
+ X6    1   3312523  862636 1132.4
+ X12   1    565679 3609480 1315.6
+ X4    1    340596 3834563 1323.4
+ X2    1     88438 4086721 1331.5
+ X5    1     77171 4097988 1331.9
<none>              4175159 1332.3
+ X9    1     39558 4135601 1333.0
+ X7    1     39324 4135835 1333.0
+ X1    1     15448 4159711 1333.8
+ X3    1     10545 4164613 1333.9
+ X10   1     10472 4164687 1333.9
+ X11   1      4176 4170982 1334.1
+ X8    1       741 4174418 1334.2

Step:  AIC=1132.41
Y ~ X6

       Df Sum of Sq     RSS     AIC
+ X4    1    584861  277775  989.36
+ X12   1    392692  469944 1056.67
+ X10   1     29736  832900 1129.92
+ X8    1     14958  847678 1132.17
<none>               862636 1132.41
+ X3    1      9152  853483 1133.05
+ X9    1      5341  857295 1133.62
+ X11   1      5119  857517 1133.65
+ X2    1       904  861732 1134.28
+ X1    1       378  862258 1134.36
+ X5    1        63  862573 1134.40
+ X7    1        36  86


Call:
lm(formula = Y ~ X6 + X4 + X12 + X10 + X2 + X11, data = data)

Coefficients:
(Intercept)           X6           X4          X12          X10           X2  
 1410.27177      5.98660      2.80806    -25.98121    -11.97829     -0.06975  
        X11  
   -0.13102  


# Hybrid yields identical results.  In the end, any of these regression models would have sufficed. Which brings to mind a question - how often does this occur, when one runs a regression model on the exact same data - do backward, forward and hybrid yield the same results often? After all, they are the same numbers. 

#### To interpret the coefficients, typically the coefficient of a variable is interpreted as the change in the response (the tests) based on a 1-unit change in the corresponding explanatory variable (IQ) keeping all other variables held constant. 